spring-cloud-stream/src/main/java/org/springframework/cloud/stream/binding/MessageConverterConfigurer.java

```diff
@@ -81,134 +80,104 @@ public void afterPropertiesSet() throws Exception {
-	 *
-	 * Will wrap the returning result of the conversion into a {@link Message} if it is not a {@link Message}
-	 * instance already.
-	 */
-	private final class MessageWrappingMessageConverter implements SmartMessageConverter {
-
-		private final MimeType contentType;
-
-		private final SmartMessageConverter delegate;
-
-		private MessageWrappingMessageConverter(SmartMessageConverter delegate, MimeType contentType) {
-			Assert.notNull(delegate, "Delegate converter cannot be null");
-			Assert.notNull(contentType, "Content type cannot be null");
-			this.delegate = delegate;
-			this.contentType = contentType;
-		}
-
-		@Override
-		public Object fromMessage(Message<?> message, Class<?> targetClass) {
-			Object converted = this.delegate.fromMessage(message, targetClass);
-			if (converted instanceof Message) {
-				return converted;
-			}
-			else {
-				return build(converted, message.getHeaders());
-			}
-		}
-
-		@Override
-		public Object fromMessage(Message<?> message, Class<?> targetClass, Object conversionHint) {
-			Object converted = this.delegate.fromMessage(message, targetClass, conversionHint);
-			if (converted == null || converted instanceof Message) {
-				return converted;
-			}
-			else {
-				return build(converted, message.getHeaders());
-			}
-		}
-
-		@Override
-		public Message<?> toMessage(Object payload, MessageHeaders headers) {
-			return this.delegate.toMessage(payload, headers);
-		}
-
-		@Override
-		public Message<?> toMessage(Object payload, MessageHeaders headers, Object conversionHint) {
-			return this.delegate.toMessage(payload, headers, conversionHint);
-		}
-
-		/**
-		 * Convenience method to construct a converted message
-		 * @param payload the converted payload
-		 * @param headers the existing message headers
-		 * @return the converted message
-		 */
-		protected Object build(Object payload, MessageHeaders headers) {
-			MimeType messageContentType = MessageConverterUtils.X_JAVA_OBJECT.equals(this.contentType) ?
-					MessageConverterUtils.javaObjectMimeType(payload.getClass()) : this.contentType;
-			return MessageConverterConfigurer.this.messageBuilderFactory.withPayload(payload).copyHeaders(headers)
-					.copyHeaders(Collections.singletonMap(MessageHeaders.CONTENT_TYPE,
-							messageContentType.toString()))
-					.build();
-		}
-	}
 
 	private final class ContentTypeConvertingInterceptor extends ChannelInterceptorAdapter {
 
 		private final String contentType;
 
 		private final MimeType mimeType;
 
-		private ContentTypeConvertingInterceptor(String contentType) {
+		private final boolean input;
+
+		private final Class<?> klazz;
+
+		private final MessageConverter messageConverter;
+
+		private ContentTypeConvertingInterceptor(String contentType, boolean input) {
 			this.contentType = contentType;
 			this.mimeType = MessageConverterUtils.getMimeType(contentType);
+			this.input = input;
+			if (MessageConverterUtils.X_JAVA_OBJECT.includes(this.mimeType)) {
```

Comment:
No, because then `application/*` would be a match. I guess we should use `equals` instead anyway.